boltrouter/bolt_info.go

```diff
@@ -0,0 +1,157 @@
+		if !castOk {
+			return nil, fmt.Errorf("could not cast availableEndpoints to []string")
+		}
+		availableEndpointsStr := make([]string, len(availableEndpointsSlice))
+		for i, v := range availableEndpointsSlice {
+			availableEndpointsStr[i] = v.(string)
+		}
+		if ok && len(availableEndpointsStr) > 0 {
+			randomIndex := rand.Intn(len(availableEndpointsStr))
+			selectedEndpoint := availableEndpointsStr[randomIndex]
+			return url.Parse(fmt.Sprintf("https://%s", selectedEndpoint))
+		}
+	}
+
+	return nil, fmt.Errorf("could not select any bolt endpoints")
+}
+
+// getPreferredEndpointOrder returns BoltVars.ReadOrderEndpoints if BoltVars.HttpReadMethodTypes contains
+// reqMethod, otherwise return BoltVars.WriteOrderEndpoints.
+func (br *BoltRouter) getPreferredEndpointOrder(reqMethod string) []string {
+	HttpReadMethodTypes := br.boltVars.HttpReadMethodTypes.Get()
+	for _, methodType := range HttpReadMethodTypes {
+		if reqMethod == methodType {
+			return br.boltVars.ReadOrderEndpoints.Get()
+		}
+	}
+
+	writeOrderEnpoints := br.boltVars.WriteOrderEndpoints.Get()
+	return writeOrderEnpoints
+}
+
+// RefreshBoltInfoPeriodically starts a goroutine that calls RefreshBoltInfo every 2min
+// TODO: refresh every 10-20 seconds
+func (br *BoltRouter) RefreshBoltInfoPeriodically(ctx context.Context) {
+	ticker := time.NewTicker(120 * time.Second)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				br.RefreshBoltInfo(ctx)
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+// RefreshBoltInfo refreshes the BoltVars.BoltInfo variable and restarts the refresh interval.
+// Call this method to force refresh BoltVars.BoltInfo.
+func (br *BoltRouter) RefreshBoltInfo(ctx context.Context) error {
+	info, err := br.getBoltInfo(ctx)
+	if err != nil {
+		return fmt.Errorf("could not refresh bolt info: %w", err)
+	}
+	br.boltVars.BoltInfo.Set(info)
+	return nil
+}
+
+// getBoltInfo queries quicksilver and returns a BoltInfo.
+// It will always return an empty mapif running in local mode.
+func (br *BoltRouter) getBoltInfo(ctx context.Context) (BoltInfo, error) {
+	// If running locally, we can't connect to quicksilver.
+	if br.config.Local {
+		return BoltInfo{}, nil
+	}
+
+	if br.boltVars.QuicksilverURL.Get() == "" || br.boltVars.Region.Get() == "" {
+		return BoltInfo{}, fmt.Errorf("quicksilverUrl: '%v' or region '%v' are not set", br.boltVars.QuicksilverURL.Get(), br.boltVars.Region.Get())
+	}
+
+	requestURL := br.boltVars.QuicksilverURL.Get()
+	r, err := http.NewRequestWithContext(ctx, http.MethodGet, requestURL, nil)
+	if err != nil {
+		return BoltInfo{}, err
+	}
+
+	resp, err := br.standardHttpClient.Do(r)
+	if err != nil {
+		return BoltInfo{}, fmt.Errorf("could not get info from quicksilver: %w", err)
+	}
+
+	defer resp.Body.Close()
+	var info BoltInfo
+	err = json.NewDecoder(resp.Body).Decode(&info)
+	if err != nil {
+		return BoltInfo{}, nil
+	}
+
+	return info, nil
```

Comment:
Can we have a log statement whenever Sidekick gets info from QS. I understand it will be a line every 10seconds. Maybe it can be DEBUG level,

Reply:
We log when the pull fails, do you think we need to log every time it succeeds?

Reply:
yes. Will be useful for debugging.

Can use log.DEBUG

Reply:
@dskart thoughts on addl logging?